torch/_higher_order_ops/wrap.py

```diff
@@ -110,24 +442,60 @@ def divide_kwargs(kwargs):
-            raise RuntimeError("Tagged Activation checkpointing does not support selective checkpointing yet.")
         import torch.fx.traceback as fx_traceback
         from torch.fx import Interpreter
-        gmod = self.tag_nodes(gmod)
-        # Using interpreter allows preservation of metadata through torch.compile stack.
-        with fx_traceback.preserve_node_meta():
-            return Interpreter(gmod).run(*args)
+        # TODO - This is a temporary sitaution where we have two versions of
+        # checkpointing implemetation. We will converge on one and remove the other.
+        if self.context_fn is not None or torch._functorch.config.functionalize_rng_ops:
+            # Mode 1: WrapActivationCheckpoint
+            assert kwargs.get("use_reentrant", False) is False, \
+                "use_reentrant=True is not supported under torch.compile " + \
+                "when selective checkpointing is used " + \
+                "or when `torch._functorch.config.functionalize_rng_ops` is True."
+            assert kwargs.get("preserve_rng_state", False) is False, \
+                "preserve_rng_state=True is not supported under torch.compile " + \
+                "when selective checkpointing is used " + \
+                "or when `torch._functorch.config.functionalize_rng_ops` is True. " + \
+                "If your program relies on this to have deterministic output in the checkpointed region, " + \
+                "it means the checkpointed region has ops that use random number generator (e.g. torch.dropout). " + \
+                "Please remove those random ops as they are not supported under selective checkpointing + torch.compile."
+
+            random_ops_in_gmod = [node.target for node in gmod.graph.nodes if node.target in random_ops]
+            assert len(random_ops_in_gmod) == 0, \
+                "Random ops are not supported in selective checkpointing region under torch.compile. " + \
+                f"Found random ops: {random_ops_in_gmod}."
+
+            inplace_ops_in_gmod = [node.target for node in gmod.graph.nodes if node.target in inplace_ops]
+            assert len(inplace_ops_in_gmod) == 0, \
+                "In-place ops are not supported in selective checkpointing region under torch.compile. " + \
+                f"Found in-place ops: {inplace_ops_in_gmod}."
+
+            log.warn("""
+Detected selective checkpointing is used under torch.compile.
+Please make sure the checkpointed region does not contain:
+
+1. random ops (e.g. torch.dropout)
+2. in-place ops (e.g. torch.relu_)
+""")
+
+            kwargs["use_reentrant"] = False
+            kwargs["preserve_rng_state"] = False
+            kwargs["context_fn"] = self.context_fn
+            # Using interpreter allows preservation of metadata through torch.compile stack.
+            with fx_traceback.preserve_node_meta():
+                return checkpoint(Interpreter(gmod).run, *args, **kwargs)
+        else:
```

Comment:
IMO, having a switch to explicitly go one way or the other is a better approach. This is with the assumption that SAC is a niche usecase.

AOT Autograd traced checkpoint requires more work around CSE etc. As that part of the stack is not hardened. I think its better if we have well tested path as default. And when that path does not work out, we emit an actionable error message to switch to the less tested path.

Reply:
CSE seems to not cause problem with selective checkpointing right now actually, but yes I can make this into a flag.

Reply:
Have you checked the memory numbers after torch.compile? CSE will silently undo AC, and will increase memory footprint.

Reply:
Note to self: check memory usage of SAC under torch.compile

Also found that `min_cut_rematerialization_partition` (which has CSE) is not used when SAC is applied, and it might explain why we don't see CSE causing problem with SAC.

Reply:
Update: I found a way to use `node.meta["recompute"]` for selective checkpointing case that's similar to full-checkpointing, details are in `CachedTorchDispatchMode._handle_compile_in_forward_ctx` method